### PR TITLE
Scripts/Ulduar: Fix Kologarn stuck in combat after death, when EVENT_SWEEP was used throughout encounter

### LINKED ISSUE #23756: Ulduar: Kologarn sometimes stuck on combat after kill
**Description:**

Kologarn sometimes stuck on combat after kill

**Current behaviour:**

Kologarn sometimes leaves party on combat after kill.
Maybe related? No achievement progress when killing it.

[00:40:00] Combat refs: (Combat state: 1 | Manager state: 1)
[00:40:00] [PvE] Armsweep Stalker Kologarn (SpawnID 136520)

**Expected behaviour:**

You exit of combat after kill it.
You get achievement progress when killing it.

**Steps to reproduce the problem:**

1. go c 136762
2. kill both arms
3. kill body
4. you stay on combat

**Branch(es):**

3.3.5

**TC rev. hash/commit:** 

4074cb0f01d1

**Operating system:** Debian 10 x64, Ubuntu 19.10 x64.

### PR DESCRIPTION
**Pull request by Artamedes**
**Changes proposed:**

-  Fix the arm sweep stalker triggers in combat after finish defeat kologarn
-  achievement criteria counting for The Antechamber of Ulduar seems to work now too

**Target branch(es):** 3.3.5/master

- [x] 3.3.5

**Issues addressed:** Closes #23756

**Tests performed:** (Does it build, tested in-game, etc.)
Builds, tested ingame

### CODE REVIEW

**Comment by Treeston** [comment]
`CombatStop` is a hack 99 times out of 100.

What are these arm sweep stalkers. Are they tempsummons? Should they be tempsummons?

**Comment by Artamedes** [comment]
They are not tempsummons, They are trigger npcs used for SPELL_ARM_SWEEP. based on sniffs, they never despawn, and are there before encounter starts

**Comment by Ovahlord** [comment]
If these sweep stalkers enter combat the spell_arm_sweep needs SPELL_ATTR0_CU_NO_INITIAL_THREAT which can be fixed easily with spell_custom_attributes.

**Comment by Killyana** [comment]
Debug combat shows 2 npc triggers:
[PvE] Armsweep Stalker Kologarn (SpawnID 136520)
[PvE] Rubble Stalker Kologarn (SpawnID 136056)

I tested the custom attribute but it looks not related to this issue:
``` sql
DELETE FROM `spell_custom_attr` WHERE `entry` IN (63766, 63983);
INSERT INTO `spell_custom_attr` (`entry`,`attributes`) VALUES
(63766,16),
(63983,16);
```

**Comment by Aokromes** [comment]
i got report also on mimiron, and

[00:53:14] Combat refs: (Combat state: 1 | Manager state: 1)
[00:53:14] [PvE] Rimefang (SpawnID 202135)

very likely something is deeper broken.

**Comment by Ovahlord** [comment]
Not really. Things like these happen mostly when a npc enters combat with you that are not suposed to enter combat. This happens withs spells that deal damage or apply some caster depending aura effect but don't have threat / aggro preventing attributes.

**Comment by Keader** [comment]
It's not a new issue, always happened.
Problem now is: after threat rewrite, triggers cant leave combat without something force it.

**Comment by jackpoz** [comment]
so what is the fix ? what is one supposed to do to spells/triggers to have TC working again ?

**Comment by Killyana** [comment]
According to sniffs:
Before the combat starts Creature Entry: 33809 (Rubble Stalker Kologarn) has this UNIT_FIELD_FLAGS: 33554432
When it cast the spell, this happen:

[2] UpdateType: Values
[2] GUID: Full: 0xF1308411000050D2 Type: Creature Entry: 33809 Low: 20690
[2] UNIT_FIELD_FLAGS: 34078720

34078720 - 33554432 =524288 (UNIT_FLAG_IN_COMBAT)

At the end of the combat UNIT_FLAG_IN_COMBAT is removed:

[0] UpdateType: Values
[0] GUID: Full: 0xF1308411000050D2 Type: Creature Entry: 33809 Low: 20690
[0] UNIT_FIELD_FLAGS: 33554432

**Comment by jackpoz** [comment]
@Treeston what should be the correct fix then ?

**Comment by jackpoz** [comment]
anyone had any update about this ?

**Comment by jackpoz** [comment]
Should the spells get SPELL_ATTR0_CU_NO_INITIAL_THREAT  ? Or what's the correct fix ? Otherwise I'll just merge this PR as it is

**Comment by jackpoz** [comment]
Any idea also about which commit added the bug ?

**Comment by Jildor** [comment]
@Keader says after @Treeston Combat/Threat Rewrite, triggers can't leave combat without something force it.

Maybe this is the main issue, I don't know if this is intended or correct or should triggers should evade after fight end like old behaviour.

**Comment by jackpoz** [comment]
is the whole point of those triggers to keep the players in combat or should they never enter combat ?

**Comment by Killyana** [comment]
According to sniffs what happen in the current revision is correct, we need just to add CombatStop() for this 2 npcs triggers when the boss reset or dies.

**Comment by Jildor** [comment]
Probably Kologarn isn't the only one with this problem, so, we need add CombatStop() to all triggers that happens similar issue?

**Comment by jackpoz** [comment]
@Artamedes could you make a reusable function that accepts creature entry (and maybe optional range) and calls combatstop to all the creatures found ? Then call that in Reset() and JustDied() and wherever it makes sense ? Thanks :)

File: src/server/scripts/Northrend/Ulduar/Ulduar/boss_kologarn.cpp [hunk 1]
```diff
@@ -152,6 +152,12 @@ class boss_kologarn : public CreatureScript
                 me->SetFlag(UNIT_FIELD_FLAGS, UNIT_FLAG_NOT_SELECTABLE);
                 me->SetCorpseDelay(604800); // Prevent corpse from despawning.
                 _JustDied();
+
+                std::list<Creature*> armSweepStalkerList;
+                me->GetCreatureListWithEntryInGrid(armSweepStalkerList, NPC_ARM_SWEEP_STALKER, 500.0f);
```
Comment (jackpoz) [inline src/server/scripts/Northrend/Ulduar/Ulduar/boss_kologarn.cpp]: is 500 needed or is 250 enough ?

**Comment by jackpoz** [comment]
tried this PR after a bit of refactoring, both arms are still in combat if you just kill the boss with gm commands.

**Comment by Killyana** [comment]
This is an other issue, both arms must die or despawn if the boss is killed.

**Comment by jackpoz** [comment]
I'll just make a PR that "Fix Kologarn stuck in combat after death" for real and apply the changes that I asked a week ago.

**Comment by jackpoz** [comment]
I created a PR with both fixes plus codestyle changes https://github.com/TrinityCore/TrinityCore/pull/23866

**Comment by Aokromes** [comment]
replaced by https://github.com/TrinityCore/TrinityCore/pull/23866

**Comment by jackpoz** [comment]
Thanks for the fix

**Pull request closed without merging**